hostmap.go

```diff
@@ -50,6 +50,13 @@ type Relay struct {
 	PeerIp      iputil.VpnIp
 }
 
+// For synchronization, treat the Relay struct as immutable. To edit the Relay
+// struct, make a copy of an existing value, edit the fileds in the copy, and
+// then store a pointer to the copy in RelayWrapper.
+type RelayWrapper struct {
```

Comment:
It looks like every time we try and get this from a map in `RelayState` we also have an `ok` check before calling `Load()`.

```
type RelayState struct {
	sync.RWMutex

	relays        map[iputil.VpnIp]struct{}
	relayForByIp  map[iputil.VpnIp]atomic.Pointer[Relay]
	relayForByIdx map[uint32]atomic.Pointer[Relay]
}

func (rs *RelayState) RemoveRelay(localIdx uint32) (iputil.VpnIp, bool) {
	rs.Lock()
	defer rs.Unlock()
	relay, ok := rs.relayForByIdx[localIdx]
	if !ok {
		return iputil.VpnIp(0), false
	}
	r := relay.Load()
	delete(rs.relayForByIdx, localIdx)
	delete(rs.relayForByIp, r.PeerIp)
	return r.PeerIp, true
}
```

Should behave the same without the additional indirection unless I have missed something.

Reply:
My motivation there was to be able to edit the Relay in-place without having to mess around with the maps. (This happens during the CompleteRelayByIp() and CompleteRelayByIdx() calls in this PR.) Since those methods have been moved to be RelayState methods in this PR, they use locks held on the RelayState struct. So, it's probably simpler/easier to maintain both maps with the lock held.

Using the atomics would require both maps use the _same_ atomic, so it'd have to be a `map[thing]*atomic.Pointer[Relay]`, but all that really gains us at this point is the ability to use a RLock in the CompleteRelayByIp|Idx() calls...which I apparently didn't do anyway, and maybe doesn't gain us a whole lot.